security/class-lockout.php

```diff
@@ -43,9 +43,9 @@ public function add_admin_notice() {
 		if ( defined( 'VIP_LOCKOUT_STATE' ) ) {
 			$user = wp_get_current_user();
 
-			switch ( VIP_LOCKOUT_STATE ) {
+			switch ( constant( 'VIP_LOCKOUT_STATE' ) ) {
```

Comment:
I don't understand why this requires the use of "constant()", usually I only see that where the name of the constant is dynamic. Can you explain for my curiosity?

Reply:
We can mock the `constant()` function, but cannot mock the constant itself.

If we use
```
define( 'CONSTANT', 'value' );
```

In a test, we have to annotate it with `@runInSeparateProcess`. PHPUnit will then run it in a separate process. But this incurs a quite significant performance penalty.

If we don't want a separate process, we need a way to "undefine" a constant somehow. But there is no way to do that on PHP (well, I can write a PHP extension for that, but that's a different story).

Because we can mock only functions, we have to use a function that will return the constant's value (`constant()`).

Now, since all our functions are namespace, PHP will look for a function in the current namespace first, and then in the global namespace.

Thus this:
```php
defined( 'VIP_JETPACK_IS_PRIVATE' ) && true === constant( 'VIP_JETPACK_IS_PRIVATE' )
```

is first tried like this:
```
\Automattic\VIP\Security\defined( 'VIP_JETPACK_IS_PRIVATE' ) && true === \Automattic\VIP\Security\constant( 'VIP_JETPACK_IS_PRIVATE' )
```

and only then as this:
```php
\defined( 'VIP_JETPACK_IS_PRIVATE' ) && true === \constant( 'VIP_JETPACK_IS_PRIVATE' )
```

Because we provide our own wrappers for [`defined()`](https://github.com/Automattic/vip-go-mu-plugins/pull/2651/files#diff-85fe9ea05584f9a0107947de85ab55ef1c121e50b020c71bc8c9de83a5367b80R38) and [`constant()`](https://github.com/Automattic/vip-go-mu-plugins/pull/2651/files#diff-85fe9ea05584f9a0107947de85ab55ef1c121e50b020c71bc8c9de83a5367b80R42), we don't actually have to define a "global" constant.

Reply:
Thanks for helping me understand this. You are a gem. 👍